### ISSUE TITLE
Cache Evictions

### ISSUE THREAD

**Issue by Feduch** [gql-dart/ferry#15]
Hi!

A few questions:
1. How can I create a forced request when the cache data is set already?
2. How I can clear cache for specific query?

**Comment by smkhalsa** [comment]
> 1. How can I create a forced request when the cache data is set already?

You can include a [`FetchPolicy`](https://github.com/gql-dart/ferry/blob/master/lib/src/client/fetch_policy.dart) when making your [`QueryRequest`](https://github.com/gql-dart/ferry/blob/master/lib/src/client/query_request.dart#L74).

> 2. How I can clear cache for specific query?

You could pass `null` to the [`Cache.writeQuery`](https://github.com/gql-dart/ferry/blob/master/lib/src/cache/cache.dart#L94) to effectively clear a specific query. However, we do not currently have any form of garbage collection, so this may leave some of the normalized data in the cache, even if no other queries reference it.

**Comment by Feduch** [comment]
Thank you!

**Comment by Feduch** [comment]
@smkhalsa I try clean cache, my code:

```
final markersGroupsQuery = markersGroups();
client.cache.writeQuery(markersGroupsQuery, null);
```
and get error:

`E/flutter (12665): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: NoSuchMethodError: The getter 'entries' was called on null.
E/flutter (12665): Receiver: null
E/flutter (12665): Tried calling: entries`

**Comment by smkhalsa** [comment]
Looks like [`this`](https://github.com/gql-dart/ferry/blob/master/lib/src/cache/cache.dart#L152) might be the issue.

We should explore evictions and garbage collection more deeply and come up with a thoughtful implementation.

**Comment by smkhalsa** [comment]
A few things that should be considered:

1. Should we remove evicted keys altogether or leave the keys with null values? I think we need to leave null values so that these values in an optimistic patch would correctly overwrite the underlying value.
1. Do we need to add the ability to evict individual fields (as Apollo does)? Or is setting null sufficient?
1. How would null values affect reading from the cache?
a. write tests to ensure that null values at both the root object and nested levels don't break reads
a. decide what to do with dangling references. Apollo [handles these](https://www.apollographql.com/docs/react/caching/garbage-collection/#dangling-references) by allowing custom read functions (which we have in normalize's FieldPolicy but haven't yet implemented). The default Apollo read function filters dangling references out of lists.
1. Investigate garbage collection methods. 
1. Do we need something similar to Apollo's retain / release methods? What if we instead create an extensible GarbageCollector class that can be customized?

**Comment by iscriptology** [comment]
Is there a way to completely invalidate the cache? (such as in a user signing-out?)

**Comment by smkhalsa** [comment]
> Is there a way to completely invalidate the cache? (such as in a user signing-out?)

Not yet, but this should be fairly trivial. All we'd need to do is clear the underlying store as well as the optimistic patch map.

**Comment by iscriptology** [comment]
I tried deleting directly the HiveStore but it didn't quite do the job.
any chance it lands soon? It causes some substantial bugs to users signing-in after past sessions.

**Comment by smkhalsa** [comment]
@iscriptology what exactly is happening when you clear the HiveStore?

**Comment by smkhalsa** [comment]
I actually see why clearing the Hive box isn't working. The [changes](https://github.com/gql-dart/ferry/blob/20dbb203432f36b768c5134302c8bd7424e679b5/ferry_hive_store/lib/src/hive_store.dart) in #63 to `HiveStore` should fix this.

Note: With the above change you will need to pass a type argument to your box, as so:

```dart
final box = await Hive.openBox<Map<String, dynamic>>('graphql');
```

**Comment by smkhalsa** [comment]
@iscriptology in v0.6.0, Cache has a `clear()` method (and so do the `Store`s. 

Please note that you need to update your store to v0.2.0.

**Comment by smkhalsa** [comment]
Please try out the new cache eviction and garbage collection methods. They've been released in ferry `0.7.2`.
